---
title: "DESeq2 Protocol"
author: "Fengyi Zheng"
date: "3/23/2023"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## **Necessary prior steps**

1. Follow [DESeq2â€™s bioconductor instruction](http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html) to download DESeq2

2. Follow [CRAN instruction](https://www.dataquest.io/blog/install-package-r/) to download other r packages

## **Helpful resource**

[DESeq2 workflow](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)


## **DESeq2**
### 1. Import packages

```{r, import packages, results = FALSE, message=FALSE, warning = FALSE}
library(data.table)
library(RColorBrewer)
library(dplyr)
library(tibble)
library(stringr)
library(readxl)
library(DESeq2)
library(ggplot2)
```


### 2. Import readcount
There are two ways to import data into DESeq2. 

1.  One is import all the experimental samples

2.  The other way is to subset the data to just the samples of interest. 


Which method you choose dependents on the specifics of the experiment. [See the disscussion on Bioconductor by the creater of DESeq2](https://support.bioconductor.org/p/p132527/). For this example, I am using two RNA-seq samples from Cas12a CRISPRa: Meljuso_RDA816_RDA810 (VP64_VP64_with_guide) and Meljuso_RDA816_RDA986 (VP64_VP64_no_guide). Each sample has three replicates.


#### 2.1 Import all readcounts in one dataframe

Raw Count Matrix is a .csv that contains raw read counts from RSEM where each column is a sample and each row is a gene.


```{r, import readcount}
base_path <- '/Users/fu/Library/CloudStorage/GoogleDrive-fu@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Fengyi/Cas12a/RNA_seq/terra/outputs/'
rsem <-'/call-rsem/'
rna_agg <-'/RNA_aggregate/call-rsem_aggregate_results/'
all_samples <- c('Meljuso_RDA763_RDA988_RepA',	'Meljuso_RDA816_RDA811_RepC', 
             'Meljuso_RDA763_RDA988_RepB',	'Meljuso_RDA816_RDA986_RepA',
             'Meljuso_RDA763_RDA988_RepC',	'Meljuso_RDA816_RDA986_RepB',
             'Meljuso_RDA816_RDA810_RepA',	'Meljuso_RDA816_RDA986_RepC',
             'Meljuso_RDA816_RDA810_RepB',	'Meljuso_RDA816_RDA987_RepA',
             'Meljuso_RDA816_RDA810_RepC',	'Meljuso_RDA816_RDA987_RepB',
             'Meljuso_RDA816_RDA811_RepA',	'Meljuso_RDA816_RDA987_RepC',
             'Meljuso_RDA816_RDA811_RepB')

# We usually use the gene results instead of isoform -- .rsem.genes.results.gz
# The aggregated isoform is equal to gene
read_files <- paste0(base_path, all_samples, rsem, all_samples, '.rsem.genes.results.gz')

reads_list <- lapply(read_files, function(x) {
  r <- fread(x, select = c("gene_id", "expected_count"))
  file_name <- unlist(strsplit(x, "Meljuso_"))[2]
  # Change the expected_count column to the sample name
  setnames(r, c("gene_id", sub("/call-rsem/", "", file_name)))
})
reads_df <- Reduce(function(x, y) merge(x, y, by = "gene_id", all=TRUE), reads_list)
reads_df<- as.data.frame(reads_df)

head(reads_df)
```

#### 2.2. You can also subset the data

```{r}
sampleid <- read_excel('/Users/fu/Library/CloudStorage/GoogleDrive-fu@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Fengyi/Cas12a/Data/rna_seq/sampleID.xlsx')
sampleid$sample_id <- sub("Meljuso_", "", sampleid$sample_id)

selected_col <- sampleid[ sampleid$Name %in%  c('VP64_VP64_with_guide', 'VP64_VP64_no_guide'),]$sample_id
selected_reads_df <- reads_df[, c("gene_id", selected_col)]
head(selected_reads_df)
```

###  3. Prepare the file to run deseq2 and select sample comparison
To successfully run DESeq2, make sure to have the row name as the gene name and numeric raw count.
```{r, prepare the file to run deseq2}
result <- data.frame(selected_reads_df, row.names = selected_reads_df$gene_id)
result <- result[,-c(1)]
result <- mutate_all(result, function(x) as.numeric(as.character(x)))
head(result)
```

### 4. Create design formula based on the samples
DESeq2 uses design formula to indicate to which samples are control and treatment. For the purpose of this example, to properly set up design formula, there should be two columns. 

- SampleID: samples name 


- Condition: indicate whether it is control or treatment aka the condition of the samples.

Note it is absolutely critical that the columns of the count matrix and the rows of the design formula (SampleID) are in the same order. 

```{r, test condition for comp_5_4 , results = TRUE, message=FALSE}
design.formula <- as.data.frame(colnames(result))
colnames(design.formula) <- 'sample_id'
design.formula <- merge(design.formula, sampleid[,c('sample_id', 'Name')], by = 'sample_id',all = FALSE)
colnames(design.formula) <- c('SampleID', 'Condition')

design.formula

```

### 5. Set up DEseq2 Data Object

DESeq2 uses *DESeqDataSetFromMatrix* to store the read counts, relationship between samples, and statistical estimates.

- countData: raw count from RSEM with all the conditions


- colData: Design Formula from step 4


- design: column name for condition in Design Formula
```{r pressure, results = FALSE, message=FALSE, warning=FALSE}

dds <- DESeqDataSetFromMatrix(countData=round(result), 
                              colData=design.formula, 
                              design=~Condition)
```

### 6. Set reference 
By default, R will choose a reference level for factors based on alphabetical order. So we need to indicate the reference level through the function *relevel*. After using the *relevel*, the first element in *Levels:* will be the new reference and the order of the data table will not change. 

Since design formula is already stored in **dds**, we can just use **dds** to move forward. To set reference, we subset the *Condition* from **dds** and set *ref* to the control indicated in the design formula. 
```{r indicate the reference level}
factor(dds$Condition)
```

```{r}
dds$Condition <- relevel(factor(dds$Condition), "VP64_VP64_no_guide")
factor(dds$Condition)
```

### 7. Pre-filtering
Remove any genes when sum of all counts across replicates and samples are less than 10. 
```{r}
dds <- dds[ rowSums(counts(dds)) > 10, ]
```

### 8. DESeq2 -- Shunken LFC 

1. DESeq: is used to establish statistical calculation from the information stored in *DESeqDataSetFromMatrix*

2. results: is used to extract a results table with log2 fold changes, p values and adjusted p values. 

3. lfcShrink: is used to perform shrunken LFC and coef is the name or number of the coefficient (LFC) to shrink.We use apeglm to shrink LFC.

4. resultsNames: is used to select the coefficient to perform shrunken LFC.

We uses shrunken LFC to measure differences in activity, as this metric allows for the shrinkage of LFC estimates towards zero for genes with low normalized read counts. 

```{r, run deseq2 , results = FALSE, message=FALSE}
dds <- DESeq(dds)
######################################################
### run those two lines if you want to see the default estimates for DESeq2
#res_dds <- results(dds) # results is an optional command to show you the estimates
#head(res_dds)
####################################
```
```{r}
resultsNames(dds)
```

```{r}
shrink_dds <- lfcShrink(dds, coef='Condition_VP64_VP64_with_guide_vs_VP64_VP64_no_guide')
shrink_dds <- as.data.frame(na.omit(shrink_dds))
head(shrink_dds)
```


### 9. Obtain normalized readcount
To extract normalized read count from DESeq2, we can the function `counts` and set normalized to True. 
```{r}
norm_count<- counts(dds, normalized=TRUE)
norm_count <- as.data.frame(norm_count)

head(norm_count)
```



## **Generate graphs**

### 1 Volcano plot
Now we can generate a volcano plot from the Shrunken LFC and adjusted p-values (DESeq2 defaults to Benjamini-Hochberg to adjust p-values)


```{r, volcano_plot, warning=FALSE}
volcano_plot <-function(deseq_input, highlight_gene = "ENSG00000010610.9", gene_name = 'CD4', plot_title = 'VP64_VP64_with_guide_vs_VP64_VP64_no_guide'){
  deseq_input <-deseq_input[complete.cases(deseq_input), ]
  deseq_input <- as.data.frame(deseq_input)
  deseq_input <- tibble::rownames_to_column(deseq_input, "ensgene")
  deseq_input[deseq_input[,'padj'] == 0,]$padj <- .Machine$double.xmin #.Machine$double.xmin gives the value of the smallest positive number whose representation meets the requirements of IEEE 754 technical standard for floating point computation.

  par(mfrow=c(1,1), cex.axis = 1.5,
  cex.lab  = 1.5, cex.main = 1.5)

  # Make a basic volcano plot
  with(deseq_input, plot(log2FoldChange, -log10(padj), pch = 1, cex=0.9, col = 'black', main=plot_title, bty = "l", xlim = c(-10,10), ylim = c(0, 310), xlab = "Shrunken Log2FoldChange")) 

  # highlighting the CD4 dot
  with(subset(deseq_input[deseq_input[,'ensgene'] == highlight_gene,]), points(log2FoldChange, -log10(padj),
      pch = 1, cex=0.9, col="red")) 

  # add CD4 text by the dot
  text(x = deseq_input[deseq_input[,'ensgene'] == highlight_gene,]$log2FoldChange , 
     y=-log10(deseq_input[deseq_input[,'ensgene'] == highlight_gene,]$padj),
     labels=gene_name, pos=4, col="red", font = 20) 

}

volcano_plot(as.data.frame(shrink_dds))
```

### 2 Baseline vs LFC plot
Now we can generate a Baseline vs LFC plot from the Shrunken LFC and baseline(control) normalized mean readcounts

```{r Baseline vs LFC plot, warning=FALSE}
bl_color_scale = c("red","#373737")
x_label = "log2(baseline expression)"
y_label = "Shrunken log2FoldChange"

# calcuate the average normalized readcount for the baseline(control)
norm_count$ctl_mean <- rowMeans(norm_count[,design.formula[design.formula$Condition == 'VP64_VP64_no_guide',]$SampleID])

# merging normalized readcount and statistical estimates from DESeq2 
together_comp <- merge(shrink_dds, norm_count, by = 0)

colnames(together_comp)[1] <- 'gene_id'

# create a column to indicate CD4
together_comp$sig <- 'False'
together_comp[together_comp[,'gene_id']  == 'ENSG00000010610.9',]$sig = 'CD4'


# graph
ggplot(together_comp %>% arrange(desc(sig)), aes(x=log2(ctl_mean), y=log2FoldChange, color=sig)) + 
  geom_point( size = 4) + 
  theme_classic() + 
  xlab(x_label) + 
  ylab(y_label) + 
  theme(text = element_text(size = 15)) +
  geom_text(aes(label=ifelse(gene_id == 'ENSG00000010610.9',as.character('CD4'),'')),hjust=0,vjust=0, size = 6)+
  scale_colour_manual(values= bl_color_scale) + ylim(-7, 10)  + guides(color = FALSE) +
  ggtitle("VP64_VP64_with_guide_vs_VP64_VP64_no_guide") +
  ggeasy::easy_center_title()

```
